bundles/framework/rpc/service/RpcService.js

```diff
@@ -0,0 +1,122 @@
+
+    /**
+     * @method create called automatically on construction
+     * @static
+     *
+     */
+
+    function (instance) {
+        this.instance = instance;
+        this.conf = this.instance.conf || {};
+        this.channel = this.instance._channel;
+        this._log = Oskari.log('RpcService');
+        this._availableFunctions = {};
+        this._allowedFunctions = this.conf.allowedfunctions || [];
+        this._allowedFunctionsFromConf = Array.isArray(this.conf.allowedfunctions);
+    }, {
+        __name: 'Rpc.RpcService',
+        __qname: 'Oskari.mapframework.bundle.rpc.service.RpcService',
+
+        getQName: function () {
+            return this.__qname;
+        },
+
+        getName: function () {
+            return this.__name;
+        },
+
+        /**
+         * @method init
+         * Initializes the service
+         */
+        init: function () {
+        },
+
+        /**
+         * @method addFunction
+         * Add function to RPC
+         *
+         * @param func function that should be added
+         */
+        addFunction: function (func) {
+            if (!func && typeof func !== 'function') {
+                return;
+            }
+
+            const me = this;
+            if (me._availableFunctions[func.name] != null) {
+                me._log.warn('Trying add already defined RPC function (' + func.name + '), skipping.');
+                return;
+            }
+
+            me._availableFunctions[func.name] = func;
+
+            // bind functions
+            me.__updateAllowedFunctions(func);
+            me.__bindFunction(func);
+        },
+
+        /**
+         * @public @method getAllowedFunctions
+         * Get allowed functions
+         *
+         * @returns {Object} allowed functions
+         */
+        getAllowedFunctions: function () {
+            return this.instance.__arrayToObject(this._allowedFunctions);
+        },
+
+        /**
+         * @private @method __bindFunction
+         * Bind allowed function to channel
+         *
+         * @param {Function} func
+         */
+        __bindFunction: function (func) {
+            const me = this;
+
+            if (!me._availableFunctions[func.name]) {
+                return;
+            }
+
+            me.channel.bind(func.name, function (trans, params) {
+                if (!me.instance._domainMatch(trans.origin)) {
+                    // eslint-disable-next-line no-throw-literal
+                    throw {
+                        error: 'invalid_origin',
+                        message: 'Invalid origin: ' + trans.origin
+                    };
+                }
+                params = params || [];
+                params.unshift(trans);
+                var value = me._availableFunctions[func.name].apply(me, params);
+                if (typeof value === 'undefined') {
+                    trans.delayReturn(true);
+                    return;
+                }
+                return value;
```

Comment:
I think we should abstract JSChannel from registered functions. This replaces transaction object from functions that _declare_ having parameters and attaches a "done" callback function instead of it to the parameters. Also try/catch the call and use transaction.error() to pass the error to the RPC-caller.

```suggestion
                const funcParams = params || [];
                const requestedFunction = me._availableFunctions[func.name];
                if (requestedFunction.length) {
                    // attach "done" function as first function parameter if it declares having any parameters
                    funcParams.unshift((response) => trans.complete(response));
                }
                try {
                    const value = requestedFunction.apply(me, funcParams);
                    if (typeof value === 'undefined') {
                        // assume it's an async function -> wait for "done()" to be called
                        trans.delayReturn(true);
                        return;
                    }
                    return value;
                } catch (err) {
                    trans.error(err, 'Error calling RPC-function: ' + func.name + 'with params: ' + params);
                }
```

This would mean that for example getScreenshot() on MapModuleClass.ol.js would need to change from:
```
        rpcService.addFunction(function getScreenshot (transaction) {
            getScreenshot(function (image) {
                transaction.complete(image);
            });
        });
```
To this:
```
        rpcService.addFunction(function getScreenshot (done) {
            getScreenshot(done);
        });
```

Reply:
Any thoughts on this?^